File: 45-jump-game-ii/jump-game-ii.cpp
class Solution {
public:
    int solve(vector<int> &nums){
        int jumps = 0, l = 0, r=0;
        int n = nums.size();
        while(r < n-1){
            int far = 0;
            for(int index = l; index <= r ;index++){
                far = max(far,index+nums[index]);
            }
            l = r + 1;
            r = far;
            jumps++;
        }

        return jumps;
    }
    int jump(vector<int>& nums) {
        int ans = solve(nums);
        return ans;
    }
};

File: 45-jump-game-ii/jump-game-ii.java
class Solution {
    public int jump(int[] nums) {
        int n = nums.length;

        if(n == 1) return 0;
        int ld = 0, end = 0, ans = 0;
        for (int i = 0; i < n; i++) {
            ld = Math.max(ld, nums[i] + i);

            if (i == end) {
                ans++;
                end = ld;
                if (ld >= n-1) {
                    return ans;
                }
            }
        }

        return ans;
    }
}